src/foundation/src/PDFsharp/src/PdfSharp/Pdf.Signatures/PdfSigner.cs

```diff
@@ -0,0 +1,257 @@
+        }
+
+        private PdfSignatureField GetOrCreateSignatureField(PdfSignatureValue value)
+        {
+            var acroForm = document.GetOrCreateAcroForm();
+            var fieldList = GetExistingFields();
+            // if a field with the specified name exist, use that
+            // Note: only root-level fields are currently supported
+            var fieldWithName = fieldList.FirstOrDefault(f => f.Name == options.FieldName);
+            if (fieldWithName != null && !(fieldWithName is PdfSignatureField))
+                throw new ArgumentException(
+                    $"Field '{options.FieldName}' exist in document, but it is not a Signature-Field ({fieldWithName.GetType().Name})");
+
+            var isNewField = false;
+            var signatureField = fieldList.FirstOrDefault(f =>
+                f is PdfSignatureField && f.Name == options.FieldName) as PdfSignatureField;
+            if (signatureField == null)
+            {
+                // field does not exist, create new one
+                signatureField = new PdfSignatureField(document)
+                {
+                    Name = options.FieldName
+                };
+                document.IrefTable.Add(signatureField);
+                acroForm.Fields.Elements.Add(signatureField);
+                isNewField = true;
+            }
+            // Flags: SignaturesExit + AppendOnly
+            acroForm.Elements.SetInteger(PdfAcroForm.Keys.SigFlags, 3);
+
+            signatureField.Value = value;
+            signatureField.Elements.SetInteger(PdfAcroField.Keys.Ff, (int)PdfAcroFieldFlags.NoExport);
+            signatureField.Elements.SetName(PdfAnnotation.Keys.Type, "/Annot");
+            signatureField.Elements.SetName(PdfAnnotation.Keys.Subtype, "/Widget");
+            if (isNewField)
+            {
+                signatureField.Elements.SetReference("/P", document.Pages[options.PageIndex]);
+                signatureField.Elements.Add(PdfAnnotation.Keys.Rect, new PdfRectangle(options.Rectangle));
+            }
+            var annotations = document.Pages[options.PageIndex].Elements.GetArray(PdfPage.Keys.Annots);
+            if (annotations == null)
+                document.Pages[options.PageIndex].Elements.Add(PdfPage.Keys.Annots, new PdfArray(document, signatureField));
+            else if (!annotations.Elements.Contains(signatureField))
+                annotations.Elements.Add(signatureField);
+
+            return signatureField;
+        }
+
+        private PdfSignatureValue CreateSignatureValue()
+        {
+            var signatureDict = new PdfSignatureValue(document);
+            document.IrefTable.Add(signatureDict);
+
+            var contentLength = GetContentLength();
+            var content = Enumerable.Repeat<byte>(0, contentLength).ToArray();
+            signatureDict.Contents = content;
+            signatureDict.Filter = "/Adobe.PPKLite";
+            signatureDict.SubFilter = "/adbe.pkcs7.detached";
```

Comment:
For my implementation I had to move these filter types public, because I had to change the subfilter to "ETSI.CAdES.detached"